tests/federation_room_join_partial_state_test.go

```diff
@@ -241,6 +242,297 @@ func TestPartialStateJoin(t *testing.T) {
+		messageId := "hiezohf6Hoo7kaev"
+		content, _ := json.Marshal(map[string]interface{}{
+			"message_id": messageId,
+			"sender":     derekUserId,
+			"type":       "m.test",
+			"messages": map[string]interface{}{
+				alice.UserID: map[string]interface{}{
+					"*": map[string]interface{}{},
+				},
+			},
+		})
+		edu := gomatrixserverlib.EDU{
+			Type:    "m.direct_to_device",
+			Content: content,
+		}
+		psjResult.Server.MustSendTransaction(t, deployment, "hs1", []json.RawMessage{}, []gomatrixserverlib.EDU{edu})
+
+		// Alice should see Derek's to-device message when she syncs.
+		alice.MustSyncUntil(t,
+			client.SyncReq{
+				Filter: buildLazyLoadingSyncFilter(nil),
+			},
+			func(userID string, sync gjson.Result) error {
+				for _, e := range sync.Get("to_device.events").Array() {
+					if e.Get("sender").Str == derekUserId &&
+						e.Get("type").Str == "m.test" {
+						return nil
+					}
+				}
+				return fmt.Errorf("No to_device update from %s", derekUserId)
+			},
+		)
+		psjResult.FinishStateRequest()
+	})
+
+	// we should be able to receive receipt EDU over federation during the resync
+	t.Run("CanReceiveReceiptDuringPartialStateJoin", func(t *testing.T) {
+		deployment := Deploy(t, b.BlueprintAlice)
+		defer deployment.Destroy(t)
+		alice := deployment.Client(t, "hs1", "@alice:hs1")
+
+		server := createTestServer(t, deployment)
+		cancel := server.Listen()
+		defer cancel()
+		serverRoom := createTestRoom(t, server, alice.GetDefaultRoomVersion(t))
+		psjResult := beginPartialStateJoin(t, server, serverRoom, alice)
+		defer psjResult.Destroy()
+
+		derekUserId := psjResult.Server.UserID("derek")
+
+		// Derek sends a read receipt into the room.
+		content, _ := json.Marshal(map[string]interface{}{
+			serverRoom.RoomID: map[string]interface{}{
+				"m.read": map[string]interface{}{
+					derekUserId: map[string]interface{}{
+						"data": map[string]interface{}{
+							"ts": 1436451550453,
+						},
+						"event_ids": []string{"mytesteventid"},
+					},
+				},
+			},
+		})
+		edu := gomatrixserverlib.EDU{
+			Type:    "m.receipt",
+			Content: content,
+		}
+		psjResult.Server.MustSendTransaction(t, deployment, "hs1", []json.RawMessage{}, []gomatrixserverlib.EDU{edu})
+
+		// Alice should be able to see Derek's read receipt during the resync
+		alice.MustSyncUntil(t,
+			client.SyncReq{
+				Filter: buildLazyLoadingSyncFilter(nil),
+			},
+			client.SyncEphemeralHas(serverRoom.RoomID, func(result gjson.Result) bool {
+				if result.Get("type").Str != "m.receipt" {
+					return false
+				}
+
+				if result.Get("content").Get("mytesteventid").Get("m\\.read").Get(strings.Replace(derekUserId, ".", "\\.", -1)).Get("ts").Int() == 1436451550453 {
+					return true
+				}
+				return false
+			}),
+		)
+		psjResult.FinishStateRequest()
+	})
+
+	// we should be able to receive device list update EDU over federation during the resync
+	t.Run("CanReceiveDeviceListUpdateDuringPartialStateJoin", func(t *testing.T) {
```

Comment:
I'm a bit dubious about this test TBH, I don't think it tests anything particularly useful as is and may just cause confusion.

Reply:
I don't fully follow this either. Shouldn't the device list update be send to Alice, even if she and Derek aren't in an encrypted room?

Reply:
As in: Alice should get told via /sync that Derek's device list has changed.

Reply:
In Synapse currently, yes, though its a bit of a grey area and we've certainly talked about only sending device list updates for E2EE enabled devices and/or users who share rooms.

OTOH, I think all the current integration tests assume this, so I don't see a problem assuming it here